pkg/cache/tas_flavor_snapshot.go

```diff
@@ -508,7 +510,8 @@ func (s *TASFlavorSnapshot) fillInCounts(requests resources.Requests,
 		if leafAssumedUsage, found := assumedUsage[leaf.domain.id]; found {
 			remainingCapacity.Sub(leafAssumedUsage)
 		}
-		leaf.state = requests.CountIn(remainingCapacity)
+		// Consider the situation where the allocatable Pods is less than allocatable resources from requests.CountIn().
+		leaf.state = min(requests.CountIn(remainingCapacity), int32(remainingCapacity[corev1.ResourcePods]))
```

Comment:
I'm onboard with @PBundyra - it is worth trying to move this logic [here](https://github.com/kubernetes-sigs/kueue/blob/06d4c763ce03891dc0f869e42d5557f9831f1682/pkg/cache/tas_flavor_snapshot.go#L301). Especially, if it helps to remove the custom checks from other places it will make the PR smaller to cherry-pick.

Reply:
Yes, that sounds reasonable. I'm trying that. Thank you for clarifying that.